gnu/usr.bin/dmidecode/man/dmidecode.8
.TH DMIDECODE 8 "February 2023" "dmidecode"
.\"
.SH NAME
dmidecode \- \s-1DMI\s0 table decoder
.\"
.SH SYNOPSIS
.B dmidecode
.RI [ OPTIONS ]
.\"
.SH DESCRIPTION
.B dmidecode
is a tool for dumping a computer's \s-1DMI\s0 (some say \s-1SMBIOS\s0) table
contents in a human-readable format. This table contains a description of the
system's hardware components, as well as other useful pieces of information
such as serial numbers and \s-1BIOS\s0 revision. Thanks to this table, you can
retrieve this information without having to probe for the actual hardware.
While this is a good point in terms of report speed and safeness, this also
makes the presented information possibly unreliable.

The \s-1DMI\s0 table doesn't only describe what the system is currently made
of, it also can report the possible evolutions (such as the fastest supported
\s-1CPU\s0 or the maximal amount of memory supported).

\s-1SMBIOS\s0 stands for System Management \s-1BIOS\s0, while \s-1DMI\s0
stands for Desktop Management Interface. Both standards are tightly related
and developed by the \s-1DMTF\s0 (Desktop Management Task Force).

As you run it,
.B dmidecode
will try to locate the \s-1DMI\s0 table. It will first try to read the DMI table
from sysfs, and next try reading directly from memory if sysfs access failed.
If
.B dmidecode
succeeds in locating a valid DMI table, it will then parse this table
and display a list of records like this one:

Handle 0x0002, DMI type 2, 8 bytes.
Base Board Information
        Manufacturer: Intel
        Product Name: C440GX+
        Version: 727281-001
        Serial Number: INCY92700942

Each record has:
.IP \(bu "\w'\(bu'u+1n"
A handle. This is a unique identifier, which allows records to
reference each other. For example, processor records usually reference
cache memory records using their handles.
.IP \(bu
A type. The \s-1SMBIOS\s0 specification defines different types of elements
a computer can be made of. In this example, the type is 2, which
means that the record contains "Base Board Information".
.IP \(bu
A size. Each record has a 4-byte header (2 for the handle, 1 for the type,
1 for the size), the rest is used by the record data. This value doesn't
take text strings into account (these are placed at the end of the record),
so the actual length of the record may be (and is often) greater than the
displayed value.
.IP \(bu
Decoded values. The information presented of course depends on the type
of record. Here, we learn about the board's manufacturer, model, version
and serial number.
.\"
.SH OPTIONS
.TP
.BR "-d" ", " "--dev-mem \fIFILE\fP"
Read memory from device \fIFILE\fP (default: \fI/dev/mem\fP)
.TP
.BR "-q" ", " "--quiet"
Be less verbose. Unknown, inactive and \s-1OEM\s0-specific entries are not
displayed. Meta-data and handle references are hidden.
.TP
.BR "  " "  " "--no-quirks"
Decode everything exactly as it is in the table, without trying to fix up
common mistakes or hide irrelevant fields.
This mode is primarily aimed at firmware developers.
.TP
.BR "-s" ", " "--string \fIKEYWORD\fP"
Only display the value of the \s-1DMI\s0 string identified by \fIKEYWORD\fP.
It must be a keyword from the following list:
.nh
.BR bios\-vendor ,
.BR bios\-version ,
.BR bios\-release\-date ,
.BR bios\-revision ,
.BR firmware\-revision ,
.BR system\-manufacturer ,
.BR system\-product\-name ,
.BR system\-version ,
.BR system\-serial\-number ,
.BR system\-uuid ,
.BR system\-sku\-number ,
.BR system\-family ,
.BR baseboard\-manufacturer ,
.BR baseboard\-product\-name ,
.BR baseboard\-version ,
.BR baseboard\-serial\-number ,
.BR baseboard\-asset\-tag ,
.BR chassis\-manufacturer ,
.BR chassis\-type ,
.BR chassis\-version ,
.BR chassis\-serial\-number ,
.BR chassis\-asset\-tag ,
.BR processor\-family ,
.BR processor\-manufacturer ,
.BR processor\-version ,
.BR processor\-frequency .
.hy
Each keyword corresponds to a given \s-1DMI\s0 type and a given offset
within this entry type.
Not all strings may be meaningful or even defined on all systems. Some
keywords may return more than one result on some systems (e.g.
.nh
.B processor\-version
.hy
on a multi-processor system).
If \fIKEYWORD\fP is not provided or not valid, a list of all valid
keywords is printed and
.B dmidecode
exits with an error.
This option cannot be used more than once.

Note: on Linux, most of these strings can alternatively be read directly
from
.BR sysfs ,
typically from files under
.IR /sys/devices/virtual/dmi/id .
Most of these files are even readable by regular users.
.TP
.BR "  " "  " "--list-strings"
List available string keywords, which can then be passed to the \fB--string\fP
option.
.TP
.BR "-t" ", " "--type \fITYPE\fP"
Only display the entries of type \fITYPE\fP. It can be either a
\s-1DMI\s0 type number, or a comma-separated list of type numbers, or a
keyword from the following list:
.nh
.BR bios ,
.BR system ,
.BR baseboard ,
.BR chassis ,
.BR processor ,
.BR memory ,
.BR cache ,
.BR connector ,
.BR slot .
.hy
Refer to the DMI TYPES section below for details.
If this option is used more than once, the set of displayed entries will be
the union of all the given types.
If \fITYPE\fP is not provided or not valid, a list of all valid keywords
is printed and
.B dmidecode
exits with an error.
.TP
.BR "  " "  " "--list-types"
List available type keywords, which can then be passed to the \fB--type\fP
option.
.TP
.BR "-H" ", " "--handle \fIHANDLE\fP"
Only display the entry whose handle matches \fIHANDLE\fP.
\fIHANDLE\fP is a 16-bit integer.
.TP
.BR "-u" ", " "--dump"
Do not decode the entries, dump their contents as hexadecimal instead.
Note that this is still a text output, no binary data will be thrown upon
you. The strings attached to each entry are displayed as both
hexadecimal and \s-1ASCII\s0. This option is mainly useful for debugging.
.TP
.BR "  " "  " "--dump-bin \fIFILE\fP"
Do not decode the entries, instead dump the DMI data to a file in binary
form. The generated file is suitable to pass to \fB--from-dump\fP
later.
\fIFILE\fP must not exist.
.TP
.BR "  " "  " "--from-dump \fIFILE\fP"
Read the DMI data from a binary file previously generated using
\fB--dump-bin\fP.
.TP
.BR "  " "  " "--no-sysfs"
Do not attempt to read DMI data from sysfs files. This is mainly useful for
debugging.
.TP
.BR "  " "  " "--oem-string \fIN\fP"
Only display the value of the \s-1OEM\s0 string number \fIN\fP. The first
\s-1OEM\s0 string has number \fB1\fP. With special value \fBcount\fP, return the
number of OEM strings instead.
.TP
.BR "-h" ", " "--help"
Display usage information and exit
.TP
.BR "-V" ", " "--version"
Display the version and exit
.P
Options
.BR --string ,
.BR --type,
.BR --dump-bin " and " --oem-string
determine the output format and are mutually exclusive.
.P
Please note in case of
.B dmidecode
is run on a system with BIOS that boasts new SMBIOS specification, which
is not supported by the tool yet, it will print out relevant message in
addition to requested data on the very top of the output. Thus informs the
output data is not reliable.
.\"
.SH "DMI TYPES"
The \s-1SMBIOS\s0 specification defines the following \s-1DMI\s0 types:
.TS
r l
__
r l.
Type	Information
0	BIOS
1	System
2	Baseboard
3	Chassis
4	Processor
5	Memory Controller
6	Memory Module
7	Cache
8	Port Connector
9	System Slots
10	On Board Devices
11	OEM Strings
12	System Configuration Options
13	BIOS Language
14	Group Associations
15	System Event Log
16	Physical Memory Array
17	Memory Device
18	32-bit Memory Error
19	Memory Array Mapped Address
20	Memory Device Mapped Address
21	Built-in Pointing Device
22	Portable Battery
23	System Reset
24	Hardware Security
25	System Power Controls
26	Voltage Probe
27	Cooling Device
28	Temperature Probe
29	Electrical Current Probe
30	Out-of-band Remote Access
31	Boot Integrity Services
32	System Boot
33	64-bit Memory Error
34	Management Device
35	Management Device Component
36	Management Device Threshold Data
37	Memory Channel
38	IPMI Device
39	Power Supply
40	Additional Information
41	Onboard Devices Extended Information
42	Management Controller Host Interface
.TE

Additionally, type 126 is used for disabled entries and type 127 is an
end-of-table marker. Types 128 to 255 are for \s-1OEM\s0-specific data.
.B dmidecode
will display these entries by default, but it can only decode them
when the vendors have contributed documentation or code for them.

Keywords can be used instead of type numbers with \fB--type\fP.
Each keyword is equivalent to a list of type numbers:

.TS
l l
__
l l.
Keyword	Types
bios	0, 13
system	1, 12, 15, 23, 32
baseboard	2, 10, 41
chassis	3
processor	4
memory	5, 6, 16, 17
cache	7
connector	8
slot	9
.TE

Keywords are matched case-insensitively. The following command lines are equivalent:
.IP \(bu "\w'\(bu'u+1n"
dmidecode --type 0 --type 13
.IP \(bu
dmidecode --type 0,13
.IP \(bu
dmidecode --type bios
.IP \(bu
dmidecode --type BIOS
.\"
.SH BINARY DUMP FILE FORMAT
The binary dump files generated by \fB--dump-bin\fP and read using \fB--from-dump\fP
are formatted as follows:
.IP \(bu "\w'\(bu'u+1n"
The SMBIOS or DMI entry point is located at offset 0x00.
It is crafted to hard-code the table address at offset 0x20.
.IP \(bu "\w'\(bu'u+1n"
The DMI table is located at offset 0x20.
.\"
.SH UUID FORMAT
There is some ambiguity about how to interpret the UUID fields prior to SMBIOS
specification version 2.6. There was no mention of byte swapping, and RFC 4122
says that no byte swapping should be applied by default. However, SMBIOS
specification version 2.6 (and later) explicitly states that the first 3 fields
of the UUID should be read as little-endian numbers (byte-swapped).
Furthermore, it implies that the same was already true for older versions of
the specification, even though it was not mentioned. In practice, many hardware
vendors were not byte-swapping the UUID. So, in order to preserve
compatibility, it was decided to interpret the UUID fields according to RFC
4122 (no byte swapping) when the SMBIOS version is older than 2.6, and to
interpret the first 3 fields as little-endian (byte-swapped) when the SMBIOS
version is 2.6 or later. The Linux kernel follows the same logic.
.\"
.SH FILES
.I /dev/mem
.br
.I /sys/firmware/dmi/tables/smbios_entry_point
(Linux only)
.br
.I /sys/firmware/dmi/tables/DMI
(Linux only)
.\"
.SH BUGS
More often than not, information contained in the \s-1DMI\s0 tables is inaccurate,
incomplete or simply wrong.
.\"
.SH AUTHORS
Alan Cox, Jean Delvare
.\"
.SH "SEE ALSO"
.BR biosdecode (8),
.BR mem (4),
.BR ownership (8),
.BR vpddecode (8)


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

overlays/uzip/furybsd-live-settings/files/usr/local/share/data/freebsd-install/dmidecode.py
# https://github.com/zaibon/py-dmidecode/
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0

import re
import subprocess


class DMIParse:
    """
    dmidecode information parsing object
    requires dmidecode output as input string
    """

    def __init__(self, str, default="n/a"):
        self.default = default
        self.data = self.dmidecode_parse(str)

    def get(self, type_id):
        if isinstance(type_id, str):
            for type_num, type_str in self.type2str.items():
                if type_str == type_id:
                    type_id = type_num

        result = list()
        for entry in self.data.values():
            if entry["DMIType"] == type_id:
                result.append(entry)
        return result

    def manufacturer(self):
        return self.get("System")[0].get("Manufacturer", self.default)

    def model(self):
        return self.get("System")[0].get("Product Name", self.default)

    def serial_number(self):
        return self.get("System")[0].get("Serial Number", self.default)

    def cpu_type(self):
        cpu_version = self.default
        for cpu in self.get("Processor"):
            if cpu.get("Core Enabled"):
                cpu_version = cpu.get("Version", self.default)
        return cpu_version

    def cpu_num(self):
        cpus = 0
        for cpu in self.get("Processor"):
            if cpu.get("Core Enabled"):
                cpus += 1
        return cpus

    def total_enabled_cores(self):
        cores = 0
        for cpu in self.get("Processor"):
            cores += int(cpu.get("Core Enabled", 0))
        return cores

    def total_ram(self):
        """Returns total memory in GB"""
        return sum([self.size_to_gb(slot["Size"]) for slot in self.get("Memory Device")])

    def firmware(self):
        return self.get("BIOS")[0].get("Firmware Revision", self.default)

    handle_re = re.compile("^Handle\\s+(.+),\\s+DMI\\s+type\\s+(\\d+),\\s+(\\d+)\\s+bytes$")
    in_block_re = re.compile("^\\t\\t(.+)$")
    record_re = re.compile("\\t(.+):\\s+(.+)$")
    record2_re = re.compile("\\t(.+):$")

    type2str = {
        0: "BIOS",
        1: "System",
        2: "Baseboard",
        3: "Chassis",
        4: "Processor",
        5: "Memory Controller",
        6: "Memory Module",
        7: "Cache",
        8: "Port Connector",
        9: "System Slots",
        10: "On Board Devices",
        11: "OEM Strings",
        12: "System Configuration Options",
        13: "BIOS Language",
        14: "Group Associations",
        15: "System Event Log",
        16: "Physical Memory Array",
        17: "Memory Device",
        18: "32-bit Memory Error",
        19: "Memory Array Mapped Address",
        20: "Memory Device Mapped Address",
        21: "Built-in Pointing Device",
        22: "Portable Battery",
        23: "System Reset",
        24: "Hardware Security",
        25: "System Power Controls",
        26: "Voltage Probe",
        27: "Cooling Device",
        28: "Temperature Probe",
        29: "Electrical Current Probe",
        30: "Out-of-band Remote Access",
        31: "Boot Integrity Services",
        32: "System Boot",
        33: "64-bit Memory Error",
        34: "Management Device",
        35: "Management Device Component",
        36: "Management Device Threshold Data",
        37: "Memory Channel",
        38: "IPMI Device",
        39: "Power Supply",
        40: "Additional Information",
        41: "Onboard Devices Extended Information",
        42: "Management Controller Host Interface",
    }

    def dmidecode_parse(self, buffer):  # noqa: C901
        data = {}
        #  Each record is separated by double newlines
        split_output = buffer.split("\n\n")

        for record in split_output:
            record_element = record.splitlines()

            #  Entries with less than 3 lines are incomplete / inactive
            #  skip them
            if len(record_element) < 3:
                continue

            handle_data = DMIDecode.handle_re.findall(record_element[0])

            if not handle_data:
                continue
            handle_data = handle_data[0]

            dmi_handle = handle_data[0]

            data[dmi_handle] = {}
            data[dmi_handle]["DMIType"] = int(handle_data[1])
            data[dmi_handle]["DMISize"] = int(handle_data[2])

            #  Okay, we know 2nd line == name
            data[dmi_handle]["DMIName"] = record_element[1]

            in_block_elemet = ""
            in_block_list = ""

            #  Loop over the rest of the record, gathering values
            for i in range(2, len(record_element), 1):
                if i >= len(record_element):
                    break
                #  Check whether we are inside a \t\t block
                if in_block_elemet != "":

                    in_block_data = DMIDecode.in_block_re.findall(record_element[1])

                    if in_block_data:
                        if not in_block_list:
                            in_block_list = in_block_data[0][0]
                        else:
                            in_block_list = in_block_list + "\t\t"
                            +in_block_data[0][1]

                        data[dmi_handle][in_block_elemet] = in_block_list
                        continue
                    else:
                        # We are out of the \t\t block; reset it again, and let
                        # the parsing continue
                        in_block_elemet = ""

                record_data = DMIDecode.record_re.findall(record_element[i])

                #  Is this the line containing handle identifier, type, size?
                if record_data:
                    data[dmi_handle][record_data[0][0]] = record_data[0][1]
                    continue

                #  Didn't findall regular entry, maybe an array of data?
                record_data2 = DMIDecode.record2_re.findall(record_element[i])

                if record_data2:
                    #  This is an array of data - let the loop know we are
                    #  inside an array block
                    in_block_elemet = record_data2[0][0]
                    continue
        return data

    def size_to_gb(self, str):
        """Convert dmidecode memory size description to GB"""
        nb = re.search("[0-9]+", str)
        if nb:
            nb = int(re.search("[0-9]+", str).group())
        else:
            return 0
        if "MB" in str:
            return nb / 1024 if nb else 0
        elif "GB" in str:
            return nb
        else:
            return 0


class DMIDecode(DMIParse):
    """Wrapper over DMIParse which runs dmidecode locally"""

    def __init__(self, command="dmidecode"):
        self.dmidecode = command
        raw = self._run()
        super().__init__(raw)

    def _run(self):
        # let subprocess merge stderr with stdout
        proc = subprocess.Popen(self.dmidecode, stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
        stdout, stderr = proc.communicate()
        if proc.returncode > 0:
            raise RuntimeError("{} failed with an error:\n{}".format(self.dmidecode, stdout.decode()))
        else:
            return stdout.decode()
